Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,362 @@
+    schedule.Push(2, start_time + chr::milliseconds(100));
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 2);
+  });
+
+  while (schedule.empty()) {
+    std::this_thread::sleep_for(chr::milliseconds(1));
+  }
+  const bool removed =
+      schedule.RemoveIf(nullptr, [](const int v) { return v == 1; });
+  EXPECT_TRUE(removed);
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(ScheduleTest, PopBlockingIsNotAffectedByIrrelevantRemovals) {
+  const auto future = std::async(std::launch::async, [&] {
+    schedule.Push(1, start_time + chr::milliseconds(50));
+    schedule.Push(2, start_time + chr::seconds(10));
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 1);
+  });
+
+  while (schedule.empty()) {
+    std::this_thread::sleep_for(chr::milliseconds(1));
+  }
+  const bool removed =
+      schedule.RemoveIf(nullptr, [](const int v) { return v == 2; });
+  EXPECT_TRUE(removed);
+  ABORT_ON_TIMEOUT(future);
+}
+
+// AsyncQueue tests
+
+namespace {
+
+class AsyncQueueTest : public ::testing::Test {
+ public:
+  AsyncQueueTest() : signal_finished{[] {}} {
+  }
+
+  // Googletest doesn't contain built-in functionality to block until an async
+  // operation completes, and there is no timeout by default. Work around both
+  // by resolving a packaged_task in the async operation and blocking on the
+  // associated future (with timeout).
+  bool WaitForTestToFinish(const chr::milliseconds timeout = kTimeout) {
+    return WaitForFuture(signal_finished.get_future(), timeout);
+  }
+
+  // Used in async tests to notify the main thread that an opration has
+  // finished.
+  std::packaged_task<void()> signal_finished;
+  AsyncQueue queue;
+};
+
+}  // namespace
+
+TEST_F(AsyncQueueTest, Enqueue) {
+  queue.Enqueue([&] { signal_finished(); });
+  EXPECT_TRUE(WaitForTestToFinish());
+}
+
+TEST_F(AsyncQueueTest, DestructorDoesNotBlockIfThereArePendingTasks) {
+  const auto future = std::async(std::launch::async, [&] {
+    AsyncQueue another_queue;
+    another_queue.EnqueueAfterDelay(chr::minutes(5), [] {});
+    another_queue.EnqueueAfterDelay(chr::minutes(10), [] {});
+    // Destructor shouldn't block waiting for the 5/10-minute-away operations.
+  });
+
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(AsyncQueueTest, CanScheduleOperationsInTheFuture) {
+  std::string steps;
+
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), [&] {
+    steps += '4';
+    signal_finished();
+  });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(1),
+                          [&steps] { steps += '3'; });
+  queue.Enqueue([&steps] { steps += '2'; });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+  EXPECT_EQ(steps, "1234");
+}
+
+TEST_F(AsyncQueueTest, CanCancelDelayedOperations) {
+  std::string steps;
+
+  queue.Enqueue([&] {
+    // Queue everything from the queue to ensure nothing completes before we
+    // cancel.
```

Comment:
That's a good idea. Would this work throughout?

Reply:
(this is a carryover from Objective-C tests)

I don't think it can help in other cases. Using `CanScheduleOperationsInTheFuture` as an example:
```cpp
  queue.Enqueue([&] {
    // omitting redundant operations
    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), [&] {
      steps += '4';
      signal_finished();
    }); // assuming `now() == 1ms`, operation '4' gets scheduled for time `6ms`
    sleep(1); // assume a context switch here
    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(1),
                            [&steps] { steps += '3'; }); // gets scheduled for `time = 1ms + 3ms + 1000ms = 1004ms`
  });
  // Expected result: "34", actual result "43"
```